libsolidity/parsing/Parser.cpp

```diff
@@ -370,6 +370,24 @@ std::pair<ContractKind, bool> Parser::parseContractKind()
 	return std::make_pair(kind, abstract);
 }
 
+ASTPointer<Expression> Parser::parseContractStorageBaseLocationExpression()
+{
+	solAssert(m_scanner->currentLiteral() == "layout");
+	expectToken(Token::Identifier);
+	if (
+		m_scanner->currentToken() != Token::Identifier ||
+		m_scanner->currentLiteral() != "at"
+	)
+		m_errorReporter.parserError(
+			1994_error,
+			m_scanner->currentLocation(),
+			"\'layout\' should be followed by \'at <storage-base-expression>\'."
+		);
+
+	advance();
+	return parseExpression();
```

Comment:
I guess it should be something more specific, because expression will accept anything.

Reply:
Yes. If I recall correctly, it must be a constant expression that can be evaluate at compile-time.  I believe we could restrict it to `TokenTraits::isArithmeticOp`, Literals and the `erc7201` helper only.

Reply:
I am assuming that we will check those kind of things later (constant, compile-time eval, etc) at type checking phase.
I am worried about parsing something weird, but I think I should look more closely to the behaviour of `parseExpression` to get a better feeling for that.